csrc/python_frontend/fusion_definition.cpp

```diff
@@ -130,14 +131,38 @@ void FusionDefinition::setupSchedule(const at::ArrayRef<c10::IValue>& inputs) {
 void FusionDefinition::finalizeSchedule(
     const at::ArrayRef<c10::IValue>& inputs) {
   FUSER_PERF_SCOPE("FusionDefinition::finalizeSchedule");
+#ifdef NVFUSER_DISTRIBUTED
+  // TODO: remove when multidevice executor integration is done natively
+  Fusion* fusion = user_sched_->schedule.get();
+  std::vector<Val*> fusion_inputs = InputsOf::outputs(fusion->outputs());
+  std::vector<Val*> vals = DependencyCheck::getAllValsBetween(
+      {fusion_inputs.begin(), fusion_inputs.end()}, fusion->outputs());
+  static Communicator* comm = new Communicator();
+  if (std::any_of(vals.begin(), vals.end(), [](Val* v) {
+        return v->isA<TensorView>() && v->as<TensorView>()->hasDeviceMesh();
+      })) {
+    multidevice_executor_ = std::make_unique<MultiDeviceExecutor>(
```

Comment:
FYI: my assumption here is that MultiDeviceExecutor wouldn't care about input "device". So there's no need to create one instances per device.

I'm actually quite curious how we'll handle this in integrated. Currently executor is fixed to one device. And support across different devices are handled by kernel_runtimes.
I imagine this would look very differently on multidevice support?

cc'ing @wujingyue @cowanmeg @samnordmann

Reply:
We are using one process per device, so yes the executor is essentially fixed to one device.
> And support across different devices are handled by kernel_runtimes.

What do you mean by this? 

> I imagine this would look very differently on multidevice support?

Like one process per node and one thread per device?

Reply:
> We are using one process per device

Wow, can you show me how this is done in code? For example, do we launch multiple bin/nvfuser_tests processes to run a multidevice test when there are multiple GPUs?

Reply:
> We are using one process per device, so yes the executor is essentially fixed to one device.

When I say `executor is fixed to one device`, I was referring to `FusionExecutor` and I was hoping that `MultiDeviceExecutor` would be agnostic for that.
Sounds like I'm totally wrong then. So I need to create a vector of `MultiDeviceExecutor` here to handle input on different devices as well...

Also, how would a broadcast look like?

i.e. input tensor is in cuda:0 and later sharded to both `cuda:0` and `cuda:1`? Do we need to feed input to both program?

Reply:
Yeah, we run all the multidevice tests with `mpirun -np <num_devices>` so nothing special  is done in the executor.

Reply:
FYI, I realize this is a bigger question and somewhat orthogonal to this PR. (given that device support on `schedule` is somewhat broken in tree).
I'm going to merge this PR without properly patching this topic.

Reply:
> So I need to create a vector of MultiDeviceExecutor here to handle input on different devices as well...

Oops I think my comment is causing confusion. No changes needed in your code since mpirun is creating the processes! 

>  input tensor is in cuda:0 and later sharded to both cuda:0 and cuda:1? Do we need to feed input to both program?

We probably need to think about how to handle sharded inputs better... Right now in our unit tests are pretty naive, we allocate the whole tensor, shard it, then pass it to the executor.

Reply:
> FYI, I realize this is a bigger question and somewhat orthogonal to this PR. (given that device support on `schedule` is somewhat broken in tree). I'm going to merge this PR without properly patching this topic.

Yes sorry, for derailing the PR. All my comments are side conversation :D

Reply:
> Oops I think my comment is causing confusion. No changes needed in your code since mpirun is creating the processes!

no I need that changes for a different case, where input changes on device across iterations. Actually I'm not totally sure if MultiDeviceExecutor supports that yet. We'll find out.

> Yes sorry, for derailing the PR.

Don't be! your comments are super helpful and I was the one asking for comments in the first place. I was just trying to communicate and clarify the scope of this PR is not going to address issues we discussed in this thread.
:bow: